sql-plugin/src/main/scala/com/nvidia/spark/rapids/SortUtils.scala

```diff
@@ -228,34 +228,79 @@ class GpuSorter(
   /**
    * Merge multiple batches together. All of these batches should be the output of
    * `appendProjectedColumns` and the output of this will also be in that same format.
-   * @param batches the batches to sort
+   *
+   * After this function is called, the argument `spillableBatches` should not be used.
+   *
+   * @param spillableBatches the spillable batches to sort
    * @param sortTime metric for the time spent doing the merge sort
    * @return the sorted data.
    */
-  final def mergeSort(batches: Array[ColumnarBatch], sortTime: GpuMetric): ColumnarBatch = {
+  final def mergeSortAndClose(
+      spillableBatches: ArrayBuffer[SpillableColumnarBatch],
+      sortTime: GpuMetric): ColumnarBatch = {
     withResource(new NvtxWithMetrics("merge sort", NvtxColor.DARK_GREEN, sortTime)) { _ =>
-      if (batches.length == 1) {
-        GpuColumnVector.incRefCounts(batches.head)
+      if (spillableBatches.length == 1) {
+        // Single batch no need for a merge sort
+        withResource(spillableBatches) { _ =>
+          spillableBatches.remove(0).getColumnarBatch()
+        }
       } else {
-        val merged = withResource(ArrayBuffer[Table]()) { tabs =>
-          batches.foreach { cb =>
-            tabs += GpuColumnVector.from(cb)
-          }
-          // In the current version of cudf merge does not work for lists and maps.
-          // This should be fixed by https://github.com/rapidsai/cudf/issues/8050
-          // Nested types in sort key columns is not supported either.
-          if (hasNestedInKeyColumns || hasUnsupportedNestedInRideColumns) {
-            // so as a work around we concatenate all of the data together and then sort it.
-            // It is slower, but it works
-            withResource(Table.concatenate(tabs: _*)) { concatenated =>
-              concatenated.orderBy(cudfOrdering: _*)
+        closeOnExcept(spillableBatches) { _ =>
+          val merged = {
+            val tablesToMerge = new ArrayBuffer[Table]()
+            while (spillableBatches.nonEmpty || tablesToMerge.size > 1) {
+              // pop a spillable batch if there is one to pop
+              var spillableCandidate: Option[SpillableColumnarBatch] = None
+              if (spillableBatches.nonEmpty) {
+                spillableCandidate = Some(spillableBatches.remove(0))
```

Comment:
Nit: An ArrayBuffer doesn't have great performance if there are many entries.  Might want to use something queue-like which also provides most of the benefits of an array, like ArrayDeque. Applies to `tablesToMerge` as well. Not must-fix.

Reply:
will look